AssertionError with extra `Array` when using `toHaveBeenCalledWith `

### Describe the bug

Hi, I get an assertion error when using Vitest with my project, porting from Vue 2 + Jest to Vue 3 + Vitest.

### Reproduction

https://github.com/poorprogrammer/cfo/commit/0a726fcc93a578cfd10d44c6201f4c50c5e2a9bf#diff-c6d94278ab82b9448cd838edb7697c8349a8c27e8f5c24daf896c50a95b72824R18

This is part of code from my senior repository.

* `presenters/__tests__/Login.spec.js`

```
import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { LoginPresenter } from "@/presenters/Login";

class MockView {
  goTo() { }
}

describe("Login Presenter", () => {
  let p = new LoginPresenter(new MockView());

  const jsdomAlert = window.alert;
  beforeEach(() => {
    window.alert = vi.fn();
  });
  afterEach(() => {
    window.alert = jsdomAlert;
  });

  /* ...... */

  it("should redirect to home page after login success", async () => {
    givenLoginSuccess();
    await p.login();
    expectToRedirectToHomePage(p.view);
  });
  /* ...... */

  function expectToRedirectToHomePage(view) {
    expect(view.goTo).toHaveBeenCalledWith({ name: "home" });
  }
});
```

* `presenters/Login.js`

```
import API from "@/services/login";
export default class {
  constructor(view) {
    this.view = view;
    this.API = new API();
  }
  login() {
    this.API.login(this.username, this.password)
      .then((token) => {
        localStorage.setItem("token", token)
        this.view.goTo({name: 'home'})
      }, (error) => {
        this.showError(error);
      });
  }
  showError(error) {
    alert(error)
  }
}
```

When run test with `npm run test`, get error like this:

```
⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Tests 1 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  src/presenters/__tests__/Login.spec.js > Login Presenter > should redirect to home page after login success
AssertionError: expected "goTo" to be called with arguments: [ { name: 'home' } ]
 ❯ expectToRedirectToHomePage src/presenters/__tests__/Login.spec.js:50:23
     48|   }
     49|   function expectToRedirectToHomePage(view) {
     50|     expect(view.goTo).toHaveBeenCalledWith({ name: "home" });
       |                       ^
     51|     // expect(view.goTo).toHaveBeenCalled();
     52|   }
 ❯ src/presenters/__tests__/Login.spec.js:29:5

  - Expected  - 3
  + Received  + 5
  
    Array [
  -   Object {
  -     "name": "home",
  -   },
  +   Array [
  +     Object {
  +       "name": "home",
  +     },
  +   ],
    ]

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯

Test Files  1 failed (1)
Tests  1 failed | 3 skipped (4)
Time  2.86s (in thread 3ms, 85531.79%)
```

**Full Repository:** https://github.com/poorprogrammer/cfo/tree/upgrade-vue/web-new

I apologize for last time with posting on duplicated issue.
If you have any feedback or suggestions to resolve this error, please let us know.

Thank you


### System Info

```shell
System:
    OS: macOS 11.5.1
    CPU: (8) arm64 Apple M1
    Memory: 82.58 MB / 8.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.13.2 - ~/.nvm/versions/node/v16.13.2/bin/node
    Yarn: 1.22.17 - ~/.nvm/versions/node/v16.13.2/bin/yarn
    npm: 8.1.2 - ~/.nvm/versions/node/v16.13.2/bin/npm
  Browsers:
    Chrome: 98.0.4758.109
    Safari: 14.1.2
  npmPackages:
    @vitejs/plugin-vue: ^2.2.2 => 2.2.2
    vite: ^2.8.4 => 2.8.4
    vitest: ^0.5.4 => 0.5.4
```


### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

It seems that the execution of `this.view.goTo({name: 'home'})` in `Login.js` is failing :thinking:

It just looks like there is an extra array because it happens to have one argument.
The reason for the object which has an extra array in the error statement is that the function may take more than one argument. 

If `goTo` is defined like
```
goTo({name: 'home'}, foo)
```
the error would  throw like this.
```
AssertionError: expected "goTo" to be called with arguments: [ { name: 'home' } , foo]
```

--- a/packages/vitest/src/integrations/chai/jest-expect.ts
+++ b/packages/vitest/src/integrations/chai/jest-expect.ts
@@ -1,11 +1,14 @@
+import p from 'picocolors'
 import type { EnhancedSpy } from '../spy'
 import { isMockFunction } from '../spy'
 import { addSerializer } from '../snapshot/port/plugins'
 import type { Constructable } from '../../types'
 import { assertTypes } from '../../utils'
+import { unifiedDiff } from '../../node/diff'
 import type { ChaiPlugin, MatcherState } from './types'
 import { arrayBufferEquality, iterableEquality, equals as jestEquals, sparseArrayEquality, subsetEquality, typeEquality } from './jest-utils'
 import type { AsymmetricMatcher } from './jest-asymmetric-matchers'
+import { stringify } from './jest-matcher-utils'
 
 const MATCHERS_OBJECT = Symbol.for('matchers-object')
 
@@ -285,6 +288,35 @@ export const JestChaiExpect: ChaiPlugin = (chai, utils) => {
     assertIsMock(assertion)
     return assertion._obj as EnhancedSpy
   }
+  const ordinalOf = (i: number) => {
+    const j = i % 10
+    const k = i % 100
+
+    if (j === 1 && k !== 11)
+      return `${i}st`
+
+    if (j === 2 && k !== 12)
+      return `${i}nd`
+
+    if (j === 3 && k !== 13)
+      return `${i}rd`
+
+    return `${i}th`
+  }
+  const formatCalls = (spy: EnhancedSpy, msg: string, actualCall?: any) => {
+    msg += `\n\nReceived: \n${spy.mock.calls.map((callArg, i) => {
+      let methodCall = p.white(`    ${ordinalOf(i + 1)} ${spy.getMockName()} call:\n\n`)
+      if (actualCall)
+        methodCall += unifiedDiff(stringify(callArg), stringify(actualCall), { showLegend: false })
+      else
+        methodCall += stringify(callArg).split('\n').map(line => `    ${line}`).join('\n')
+
+      methodCall += '\n'
+      return methodCall
+    }).join('\n')}`
+    msg += `\n\nNumber of calls: ${p.bold(spy.mock.calls.length)}\n`
+    return msg
+  }
   def(['toHaveBeenCalledTimes', 'toBeCalledTimes'], function(number: number) {
     const spy = getSpy(this)
     const spyName = spy.getMockName()
@@ -313,41 +345,49 @@ export const JestChaiExpect: ChaiPlugin = (chai, utils) => {
     const spy = getSpy(this)
     const spyName = spy.getMockName()
     const called = spy.mock.calls.length > 0
-    return this.assert(
-      called,
-      `expected "${spyName}" to be called at least once`,
-      `expected "${spyName}" to not be called at all`,
-      true,
-      called,
+    const isNot = utils.flag(this, 'negate') as boolean
+    let msg = utils.getMessage(
+      this,
+      [
+        called,
+        `expected "${spyName}" to be called at least once`,
+        `expected "${spyName}" to not be called at all`,
+        true,
+        called,
+      ],
     )
+    if (called && isNot)
+      msg += formatCalls(spy, msg)
+
+    if ((called && isNot) || (!called && !isNot)) {
+      const err = new Error(msg)
+      err.name = 'AssertionError'
+      throw err
+    }
   })
   def(['toHaveBeenCalledWith', 'toBeCalledWith'], function(...args) {
     const spy = getSpy(this)
     const spyName = spy.getMockName()
     const pass = spy.mock.calls.some(callArg => jestEquals(callArg, args, [iterableEquality]))
-    return this.assert(
-      pass,
-      `expected "${spyName}" to be called with arguments: #{exp}`,
-      `expected "${spyName}" to not be called with arguments: #{exp}`,
-      args,
-      spy.mock.calls,
+    const isNot = utils.flag(this, 'negate') as boolean
+
+    let msg = utils.getMessage(
+      this,
+      [
+        pass,
+        `expected "${spyName}" to be called with arguments: #{exp}`,
+        `expected "${spyName}" to not be called with arguments: #{exp}`,
+        args,
+      ],
     )
+
+    if ((pass && isNot) || (!pass && !isNot)) {
+      msg += formatCalls(spy, msg, args)
+      const err = new Error(msg)
+      err.name = 'AssertionError'
+      throw err
+    }
   })
-  const ordinalOf = (i: number) => {
-    const j = i % 10
-    const k = i % 100
-
-    if (j === 1 && k !== 11)
-      return `${i}st`
-
-    if (j === 2 && k !== 12)
-      return `${i}nd`
-
-    if (j === 3 && k !== 13)
-      return `${i}rd`
-
-    return `${i}th`
-  }
   def(['toHaveBeenNthCalledWith', 'nthCalledWith'], function(times: number, ...args: any[]) {
     const spy = getSpy(this)
     const spyName = spy.getMockName()

--- a/packages/vitest/src/integrations/chai/jest-expect.ts
+++ b/packages/vitest/src/integrations/chai/jest-expect.ts
@@ -1,4 +1,4 @@
-import p from 'picocolors'
+import c from 'picocolors'
 import type { EnhancedSpy } from '../spy'
 import { isMockFunction } from '../spy'
 import { addSerializer } from '../snapshot/port/plugins'
@@ -304,8 +304,8 @@ export const JestChaiExpect: ChaiPlugin = (chai, utils) => {
     return `${i}th`
   }
   const formatCalls = (spy: EnhancedSpy, msg: string, actualCall?: any) => {
-    msg += `\n\nReceived: \n${spy.mock.calls.map((callArg, i) => {
-      let methodCall = p.white(`    ${ordinalOf(i + 1)} ${spy.getMockName()} call:\n\n`)
+    msg += c.gray(`\n\nReceived: \n${spy.mock.calls.map((callArg, i) => {
+      let methodCall = c.bold(`    ${ordinalOf(i + 1)} ${spy.getMockName()} call:\n\n`)
       if (actualCall)
         methodCall += unifiedDiff(stringify(callArg), stringify(actualCall), { showLegend: false })
       else
@@ -313,8 +313,8 @@ export const JestChaiExpect: ChaiPlugin = (chai, utils) => {
 
       methodCall += '\n'
       return methodCall
-    }).join('\n')}`
-    msg += `\n\nNumber of calls: ${p.bold(spy.mock.calls.length)}\n`
+    }).join('\n')}`)
+    msg += c.gray(`\n\nNumber of calls: ${c.bold(spy.mock.calls.length)}\n`)
     return msg
   }
   def(['toHaveBeenCalledTimes', 'toBeCalledTimes'], function(number: number) {

--- a/packages/vitest/src/integrations/chai/jest-matcher-utils.ts
+++ b/packages/vitest/src/integrations/chai/jest-matcher-utils.ts
@@ -3,6 +3,7 @@
 
 import c from 'picocolors'
 import type { Formatter } from 'picocolors/types'
+import type { PrettyFormatOptions } from 'pretty-format'
 import { format as prettyFormat, plugins as prettyFormatPlugins } from 'pretty-format'
 import { unifiedDiff } from '../../node/diff'
 
@@ -112,7 +113,7 @@ const SPACE_SYMBOL = '\u{00B7}' // middle dot
 const replaceTrailingSpaces = (text: string): string =>
   text.replace(/\s+$/gm, spaces => SPACE_SYMBOL.repeat(spaces.length))
 
-export const stringify = (object: unknown, maxDepth = 10): string => {
+export const stringify = (object: unknown, maxDepth = 10, options?: PrettyFormatOptions): string => {
   const MAX_LENGTH = 10000
   let result
 
@@ -121,6 +122,7 @@ export const stringify = (object: unknown, maxDepth = 10): string => {
       maxDepth,
       // min: true,
       plugins: PLUGINS,
+      ...options,
     })
   }
   catch {
@@ -129,6 +131,7 @@ export const stringify = (object: unknown, maxDepth = 10): string => {
       maxDepth,
       // min: true,
       plugins: PLUGINS,
+      ...options,
     })
   }
 

--- a/packages/vitest/src/node/diff.ts
+++ b/packages/vitest/src/node/diff.ts
@@ -6,6 +6,11 @@ export function formatLine(line: string, outputTruncateLength?: number) {
   return cliTruncate(line, (outputTruncateLength ?? (process.stdout.columns || 80)) - 4)
 }
 
+export interface DiffOptions {
+  outputTruncateLength?: number
+  showLegend?: boolean
+}
+
 /**
 * Returns unified diff between two strings with coloured ANSI output.
 *
@@ -15,9 +20,11 @@ export function formatLine(line: string, outputTruncateLength?: number) {
 * @return {string} The diff.
 */
 
-export function unifiedDiff(actual: string, expected: string, outputTruncateLength?: number) {
+export function unifiedDiff(actual: string, expected: string, options: DiffOptions = {}) {
   if (actual === expected)
     return ''
+
+  const { outputTruncateLength, showLegend = true } = options
 
   const indent = '  '
   const diffLimit = 15
@@ -70,19 +77,21 @@ export function unifiedDiff(actual: string, expected: string, outputTruncateLeng
     return ` ${line}`
   })
 
-  // Compact mode
-  if (isCompact) {
-    formatted = [
-      `${c.green('- Expected')}   ${formatted[0]}`,
-      `${c.red('+ Received')}   ${formatted[1]}`,
-    ]
-  }
-  else {
-    formatted.unshift(
-      c.green(`- Expected  - ${counts['-']}`),
-      c.red(`+ Received  + ${counts['+']}`),
-      '',
-    )
+  if (showLegend) {
+    // Compact mode
+    if (isCompact) {
+      formatted = [
+        `${c.green('- Expected')}   ${formatted[0]}`,
+        `${c.red('+ Received')}   ${formatted[1]}`,
+      ]
+    }
+    else {
+      formatted.unshift(
+        c.green(`- Expected  - ${counts['-']}`),
+        c.red(`+ Received  + ${counts['+']}`),
+        '',
+      )
+    }
   }
 
   return formatted.map(i => indent + i).join('\n')

--- a/packages/vitest/src/node/error.ts
+++ b/packages/vitest/src/node/error.ts
@@ -86,7 +86,7 @@ function handleImportOutsideModuleError(stack: string, ctx: Vitest) {
 }
 
 function displayDiff(actual: string, expected: string, console: Console, outputTruncateLength?: number) {
-  console.error(c.gray(unifiedDiff(actual, expected, outputTruncateLength)) + '\n')
+  console.error(c.gray(unifiedDiff(actual, expected, { outputTruncateLength })) + '\n')
 }
 
 function printErrorMessage(error: ErrorWithDiff, console: Console) {

